presto-blackhole/src/main/java/com/facebook/presto/plugin/blackhole/BlackHolePageSourceProvider.java

```diff
@@ -19,18 +19,38 @@
 import com.facebook.presto.spi.ConnectorPageSourceProvider;
 import com.facebook.presto.spi.ConnectorSession;
 import com.facebook.presto.spi.ConnectorSplit;
-import com.facebook.presto.spi.InMemoryRecordSet;
+import com.facebook.presto.spi.Page;
 import com.facebook.presto.spi.RecordPageSource;
+import com.facebook.presto.spi.type.Type;
 import com.google.common.collect.ImmutableList;
 
 import java.util.List;
 
+import static com.facebook.presto.plugin.blackhole.Types.checkType;
+
 public final class BlackHolePageSourceProvider
         implements ConnectorPageSourceProvider
 {
     @Override
     public ConnectorPageSource createPageSource(ConnectorSession session, ConnectorSplit split, List<ColumnHandle> columns)
     {
-        return new RecordPageSource(new InMemoryRecordSet(ImmutableList.of(), ImmutableList.of()));
+        BlackHoleSplit blackHoleSplit = checkType(split, BlackHoleSplit.class, "BlackHoleSplit");
+
+        ImmutableList.Builder<Type> builder = ImmutableList.builder();
+
+        for (ColumnHandle column : columns) {
+            builder.add((checkType(column, BlackHoleColumnHandle.class, "BlackHoleColumnHandle")).getColumnType());
+        }
+        List<Type> types = builder.build();
+
+        return new ConstantPageSource(
+                generateZeroPage(types, blackHoleSplit.getRowsPerPage()),
+                blackHoleSplit.getPagesCount());
+    }
+
+    private Page generateZeroPage(List<Type> types, int rowsCount)
+    {
+        ConnectorPageSource pageSource = new RecordPageSource(types, new BlackHoleRecordCursor(types, rowsCount));
```

Comment:
Not much harder. You only need to create a Page.

Incidentally, the implementation of BlackHoleRecordCursor is broken (I just realized this). The getXXX methods return a default value which may not be a valid value for the corresponding type. For instance, a SQL timestamp may be encoded as a long, but "0" may be an invalid encoding. We're probably going to extend types so that they can provide a "default" value.